### PR TITLE
706 slow dns resolutions

### PR DESCRIPTION
**Pull request by mitesh-gosavi**
**Description of changes**

Mplotqueries along with —dns flag has been used to search for log lines containing the “DNS resolution while connecting to” token (For reference, see https://github.com/mongodb/mongo/commit/04806cf67c2d2d738b996183eb2c733e23b2afae). On encountering the token, the parser parses the token after “took” and typecast it to an integer to get the duration.
The connector host has parsed by parsing the token between “to” and “took” in the log line.

**Testing**

The following testing tools have been used to test the newly added code:
1. Tox
2. Flake8
3. iSort
4. Doc8

The test result of a code coverage can be seen in table below:

**Name                                              Stmts   Miss  Cover
---------------------------------------------------------------------**

mtools/mplotqueries/mplotqueries.py                 400     90    78%
mtools/util/logevent.py                             514      2    99%

**Version Information**

| macOS 		| 10.14.5	|
| mongod		| v4.0.10	|
| MongoDB shell	| v4.0.10	|
| Python		| 2.7.10		|

### CODE REVIEW

File: mtools/mplotqueries/mplotqueries.py [hunk 1]
```diff
@@ -81,7 +84,7 @@ def __init__(self):
                                     help=("specify value to group on. "
                                           "Possible values depend on type of "
                                           "plot. All basic plot types can "
-                                          "group on 'namespace', 'operation', "
+                                          "group on 'namespace','connector' 'operation', "
```
Comment (HenryGP) [inline mtools/mplotqueries/mplotqueries.py]: What do you mean by _connector_ here?

File: mtools/util/logevent.py [hunk 2]
```diff
@@ -205,6 +207,16 @@ def duration(self):
 
         return self._duration
 
+    #SERVER-41349 - get connector from the DNS log line
```
Comment (HenryGP) [inline mtools/util/logevent.py]: connector could be interpreted as something else in this context 🤔

**Review by HenryGP — CHANGES_REQUESTED** [review]
Small wording changes for the comments and help.

File: mtools/mplotqueries/mplotqueries.py (hunk 1 above)
Comment (savinay-vijay) [reply mtools/mplotqueries/mplotqueries.py]: Agreed. It's meant to be hostname. Improper terminology there from me. I'll update that to hostname

File: mtools/mplotqueries/mplotqueries.py [hunk 3]
```diff
@@ -69,6 +70,8 @@ def __init__(self):
                                           "all existing overlays. Use "
                                           "'--overlay reset' to clear all "
                                           "overlays."))
+        # SERVER-41349 - --dns flag to plot slow DNS Resolutions; Can be grouped by connector
```
Comment (davidedeangelismdb) [inline mtools/mplotqueries/mplotqueries.py]: as per Enrique comment, same here about connector. Remove the SERVER ticket reference.

File: mtools/mplotqueries/mplotqueries.py [hunk 4]
```diff
@@ -198,6 +201,9 @@ def parse_logevents(self):
                     self.progress_bar_enabled = False
 
             for i, logevent in enumerate(logfile):
+                if (self.args['dns'] and (
```
Comment (davidedeangelismdb) [inline mtools/mplotqueries/mplotqueries.py:204]: is the key 'dns' always part of the args passed to the tool? what if is not there? is there a default value?

File: mtools/util/logevent.py (hunk 2 above)
Comment (davidedeangelismdb) [reply mtools/util/logevent.py]: same above remove the ticket reference

**Review by davidedeangelismdb — CHANGES_REQUESTED** [review]

File: mtools/mplotqueries/mplotqueries.py (hunk 4 above)
Comment (savinay-vijay) [reply mtools/mplotqueries/mplotqueries.py:204]: --dns acts as a filter to parse (and plot) only lines containing DNS log entries. if --dns argument is not provided then all log lines will be parsed and plotted. by default, the parser plots all lines containing an operation or command.

File: mtools/util/logevent.py (hunk 2 above)
Comment (HenryGP) [reply mtools/util/logevent.py]: Well, maybe the SERVER ticket reference is ok. It is a public ticket anyway and provides further context about what the intention is for the functionality added.

File: mtools/mplotqueries/mplotqueries.py (hunk 3 above)
Comment (thiagogalesi) [reply mtools/mplotqueries/mplotqueries.py]: seconded

File: mtools/test/test_mplotqueries.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import sys
+from datetime import datetime, timedelta
+from random import randrange
+
+from dateutil import parser
+from nose.plugins.skip import SkipTest
+from nose.tools import raises
+
+import os
```
Comment (thiagogalesi) [inline mtools/test/test_mplotqueries.py:9]: you can put the builtin imports up there with the other builtins (usually they're order alphabetically)

File: mtools/util/logevent.py [hunk 6]
```diff
@@ -205,6 +207,16 @@ def duration(self):
...
+    @property
+    def connector(self):
+        line_str = self.line_str
+        if(line_str and 'DNS resolution' in line_str):
+            searchString = "connecting to"
+            self._connector = line_str[line_str.rfind(searchString) + len(searchString) + 1:-12]
```
Comment (thiagogalesi) [inline mtools/util/logevent.py]: that works but it is not nice and really hard to read :) 

You have used regexes in another part of the code but you can use them here to make it easier

Using () denotes groups and the search will give it directly https://docs.python.org/3/howto/regex.html

try this (example that you'll have to adapt):

```
>>> a = re.search("DNS resolution while connecting to ([\w.]+) took ([\d]+)ms", line_str)
>>> a
<re.Match object; span=(10, 60), match='DNS resolution while connecting to ABC took 5000m>
>>> a.groups()
('ABC', '5000')
```

**Review by thiagogalesi — COMMENTED** [review]
some comments

File: mtools/mplotqueries/mplotqueries.py (hunk 1 above)
Comment (manfontan) [reply mtools/mplotqueries/mplotqueries.py]: Is there a missing comma before operations? My understanding id that you can name by `namespace`, `hostname`, `operation`

File: mtools/test/test_mplotqueries.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import mtools
+from mtools.mplotqueries.mplotqueries import MPlotQueriesTool
+from mtools.util.logevent import LogEvent
+from mtools.util.logfile import LogFile
+
+
+class TestMPlotQueries(object):
+
+    def setup(self):
+        """Startup method to create mloginfo tool."""
```
Comment (manfontan) [inline mtools/test/test_mplotqueries.py]: mloginfo? shouldn't it be mplotqueries ?

Comment (mitesh-gosavi) [reply mtools/test/test_mplotqueries.py]: Yes, duly noted in the list of changes to be made.

File: mtools/mplotqueries/mplotqueries.py [hunk 8]
```diff
@@ -19,6 +19,7 @@
     from matplotlib.dates import AutoDateFormatter, date2num, AutoDateLocator
     from matplotlib import __version__ as mpl_version
     import mtools.mplotqueries.plottypes as plottypes
+    import re
```
Comment (manfontan) [inline mtools/mplotqueries/mplotqueries.py:22]: Review PEP style guide to sort imports https://www.python.org/dev/peps/pep-0008/#imports ;)
```suggestion
    import re
```

File: mtools/test/test_mplotqueries.py (hunk 5 above)
Comment (manfontan) [reply mtools/test/test_mplotqueries.py:9]: As per Thiagho's comment review PEP style guide to sort imports https://www.python.org/dev/peps/pep-0008/#imports ;)

**Review by HenryGP — APPROVED** [review]

**Review by davidedeangelismdb — APPROVED** [review]

File: mtools/util/logevent.py [hunk 9]
```diff
@@ -205,6 +207,14 @@ def duration(self):
 
         return self._duration
 
+    #SERVER-41349 - get hostname from the DNS log line
+    @property
```
Comment (thiagogalesi) [inline mtools/util/logevent.py:222]: 👍

**Review by thiagogalesi — APPROVED** [review]
Ok LGTM!

**Pull request merged**